clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1260,6 +1283,41 @@ class SyclKernelBodyCreator
     InitExprs.push_back(MemberInit.get());
   }
 
+  void createExprForArray(FieldDecl *FD) {
+    ParmVarDecl *KernelParameter =
+        DeclCreator.getParamVarDeclsForCurrentField()[0];
+    QualType ParamType = KernelParameter->getOriginalType();
+    CXXRecordDecl *WrapperStruct = ParamType->getAsCXXRecordDecl();
+    // The first and only field of the wrapper struct is the array
+    FieldDecl *Array = *(WrapperStruct->field_begin());
+    auto DRE = DeclRefExpr::Create(SemaRef.Context, NestedNameSpecifierLoc(),
+                                   SourceLocation(), KernelParameter, false,
+                                   DeclarationNameInfo(), ParamType, VK_LValue);
+    DeclAccessPair ArrayDAP = DeclAccessPair::make(Array, AS_none);
+    Expr *InitExpr = MemberExpr::Create(
+        SemaRef.Context, DRE, false, SourceLocation(), NestedNameSpecifierLoc(),
+        SourceLocation(), Array, ArrayDAP,
+        DeclarationNameInfo(Array->getDeclName(), SourceLocation()), nullptr,
+        Array->getType(), VK_LValue, OK_Ordinary, NOUR_None);
+    InitializationKind InitKind = InitializationKind::CreateDirect(
+        SourceLocation(), SourceLocation(), SourceLocation());
+    InitializedEntity Entity = InitializedEntity::InitializeLambdaCapture(
+        nullptr, Array->getType(), SourceLocation());
```

Comment:
This does bulk initialization of an array. Lambda capture of an array uses this pattern (outside a SYCL context), so the same pattern was used here.

Reply:
Will case with kernel represented by a function object be handled properly?

Reply:
Yes, it works. When a function object is used as a kernel parameter and the object contains an initialized array, the array is received on the device with its initialization copied over.

Reply:
I'd think we would want to not depend on the lambda capture, since the structure of a lambda object isn't stable.  In fact, there was an effort recently to change the ordering of members in an optimization attempt on llvm-dev a few months ago.

It would seem that InitializeMember fits better.  More correctly, there is InitializeElement that handles array elements.